File: app/android/app/src/main/kotlin/me/jxl/kiosk_satellite/AppIdentity.kt
package me.jxl.kiosk_satellite

import android.content.Context
import android.os.Build

/**
 * How Kiosk Satellite introduces itself on the wire from the native side.
 *
 * The Dart half has its own copy of this string (core/app_identity.dart)
 * and hands it to every dart:io client; the players and sockets that live
 * in Kotlin have their own HTTP stacks (OkHttp, Media3) that would
 * otherwise announce themselves as the library rather than the app. Same
 * shape either way, so one log line reads the same wherever it came from.
 *
 * [configure] runs from the Activity, which is the only place with a
 * Context guaranteed to be up before anything connects; callers deep in
 * the audio stack just read [userAgent], which falls back to the bare
 * product name if it is somehow read first.
 */
object AppIdentity {
    private const val PRODUCT = "KioskSatellite"
    private const val HOMEPAGE = "https://github.com/jxlarrea/kiosk-satellite"

    @Volatile
    private var cached: String? = null

    /** `KioskSatellite/2026.8.14 (Android 12; +https://github.com/...)`. */
    val userAgent: String get() = cached ?: PRODUCT

    /**
     * No device model in the string: it reaches servers the app does not
     * own, and the model is a fingerprint they have no business collecting.
     */
    fun configure(context: Context) {
        if (cached != null) return
        val version = try {
            context.packageManager
                .getPackageInfo(context.packageName, 0).versionName ?: "unknown"
        } catch (_: Exception) {
            "unknown"
        }
        cached = "$PRODUCT/$version (Android ${Build.VERSION.RELEASE}; +$HOMEPAGE)"
    }
}

File: app/lib/core/app_identity.dart
/// How Kiosk Satellite introduces itself on the wire.
///
/// Every request the app makes from Dart used to go out as
/// `Dart/3.9 (dart:io)`, which tells the person reading a Go2RTC, Home
/// Assistant or reverse-proxy log nothing at all: not the app, not its
/// version, not even that the traffic came from a kiosk. One string, set
/// on the global [HttpOverrides] client, covers every HTTP request and
/// every websocket the app opens.
///
/// The WebView keeps its own browser user agent: pages are browsed as a
/// browser, and Home Assistant's frontend reads that string.
class AppIdentity {
  const AppIdentity._();

  static const product = 'KioskSatellite';
  static const homepage = 'https://github.com/jxlarrea/kiosk-satellite';

  /// The user agent every outbound request carries. Set once at startup;
  /// until then it is the bare product name, so a request that somehow
  /// beats [configure] still identifies the app.
  static String userAgent = product;

  /// `KioskSatellite/2026.8.14 (Android 12; +https://github.com/...)`.
  /// The version answers "which build is doing this", the OS answers the
  /// first question every support thread asks, and the link is what
  /// services that require a contactable user agent (LRCLIB) ask for. No
  /// device model: this string reaches third-party servers, and the model
  /// is a fingerprint they have no business collecting.
  static void configure({required String version, required String osVersion}) {
    final os = osVersion.trim();
    userAgent =
        '$product/$version (${os.isEmpty ? '' : '$os; '}+$homepage)';
  }
}
